testng-core/src/test/java/test/reports/FailedReporterTest.java

```diff
@@ -81,4 +90,123 @@ public void testParameterPreservationWithFactory() throws IOException {
+        .withFailMessage("The failed test should have had the suffix of [(failed)]")
+        .isEqualTo("kungfu-panda-test(failed)");
+    runIncludedMethodsAssertion(
+        failedTest.getClasses(), pairA.getTestClass(), pairA.getTestMethods());
+    runIncludedMethodsAssertion(
+        failedTest.getClasses(), pairB.getTestClass(), pairB.getTestMethods());
+  }
+
+  @DataProvider(name = "getTestData")
+  public Object[][] getTestData() {
+    return new Object[][] {
+      {
+        new ClassMethodInfo(TestClassWithBeforeTestSample.class, "beforeTest", "afterTest"),
+        new ClassMethodInfo(TestClassWithJustTestMethodsSample.class, "test2Method")
+      },
+      {
+        new ClassMethodInfo(TestClassWithBeforeSuiteSample.class, "beforeSuite", "afterSuite"),
+        new ClassMethodInfo(TestClassWithJustTestMethodsSample.class, "test2Method")
+      },
+      {
+        new ClassMethodInfo(TestClassWithBeforeGroupsSample.class, "beforeGroups", "afterGroups"),
+        new ClassMethodInfo(TestClassWithJustTestMethodsSample.class, "test2Method")
+      },
+      {
+        new ClassMethodInfo(
+            TestClassFailsAtBeforeGroupsWithBeforeGroupsSuiteTestSample.class,
+            "beforeGroups",
+            "afterGroups",
+            "beforeTest",
+            "afterTest",
+            "beforeSuite",
+            "afterSuite"),
+        new ClassMethodInfo(TestClassWithJustTestMethodsSample.class, "test2Method")
+      },
+      {
+        new ClassMethodInfo(
+            TestClassFailsAtBeforeSuiteWithBeforeGroupsSuiteTestSample.class,
+            "beforeGroups",
+            "afterGroups",
+            "beforeTest",
+            "afterTest",
+            "beforeSuite",
+            "afterSuite"),
+        new ClassMethodInfo(TestClassWithJustTestMethodsSample.class, "test2Method")
+      },
+      {
+        new ClassMethodInfo(
+            TestClassFailsAtBeforeTestWithBeforeGroupsSuiteTestSample.class,
+            "beforeGroups",
+            "afterGroups",
+            "beforeTest",
+            "afterTest",
+            "beforeSuite",
+            "afterSuite"),
+        new ClassMethodInfo(TestClassWithJustTestMethodsSample.class, "test2Method")
+      }
+    };
+  }
+
+  private static void runIncludedMethodsAssertion(
+      List<XmlClass> failedClasses, Class<?> cls, String... methods) {
+    XmlClass xmlClass =
+        failedClasses.stream()
+            .filter(each -> each.getName().equals(cls.getName()))
+            .findFirst()
+            .orElseThrow(() -> new IllegalStateException("Failed to locate " + cls.getName()));
+    List<String> includedMethods =
+        xmlClass.getIncludedMethods().stream()
+            .map(XmlInclude::getName)
+            .collect(Collectors.toList());
+    assertThat(includedMethods)
+        .withFailMessage("Included methods test failed for " + cls.getName())
+        .containsExactlyInAnyOrder(methods);
+  }
+
+  public static class ClassMethodInfo {
```

Comment:
WDYT to use https://projectlombok.org/ in test to avoid boilerplate code?

Reply:
@juherr - Yes that's a very good option. But I dont know if we have enough number needs in tests to take this route.